db/compaction_picker.cc

```diff
@@ -661,30 +661,29 @@ Compaction* CompactionPicker::CompactRange(
   }
   assert(output_path_id < static_cast<uint32_t>(ioptions_.cf_paths.size()));
 
-  // for BOTTOM LEVEL compaction, skip files that are created during the
-  // current compaction
-  if (max_sst_file_number > 0) {
+  // for BOTTOM LEVEL compaction only, use next_file_number to filter out files
+  // that are created during the current compaction.
+  if (next_file_number != port::kMaxUint64) {
     assert(input_level == output_level);
-    std::vector<FileMetaData*> inputs_shrinked;
+    // inputs_shrunk holds a continuous subset of input files which were all
+    // created before the current manual compaction
+    std::vector<FileMetaData*> inputs_shrunk;
     for (size_t i = 0; i < inputs.size(); ++i) {
-      if (inputs[i]->fd.GetNumber() <= max_sst_file_number) {
-        inputs_shrinked.emplace_back(inputs[i]);
-      } else {
+      if (inputs[i]->fd.GetNumber() < next_file_number) {
+        inputs_shrunk.push_back(inputs[i]);
+      } else if (!inputs_shrunk.empty()) {
         // inputs[i] was created during the current manual compaction and
         // need to be skipped
-        if (inputs_shrinked.empty()) {
-          // haven't found anything to compact, keep looking
-          continue;
-        } else {
-          // already found something to compact, skip the rest files
-          break;
-        }
+        break;
       }
     }
-    if (inputs_shrinked.empty()) {
+    if (inputs_shrunk.empty()) {
       return nullptr;
     }
-    inputs.files.swap(inputs_shrinked);
+    if (inputs.size() != inputs_shrunk.size()) {
+      inputs.files.swap(inputs_shrunk);
+      covering_the_whole_range = false;
```

Comment:
Why?

Reply:
because `inputs` is shrunk here, so this compaction no longer covers the whole range. the check for `max_compaction_bytes` seems to be doing the same thing, no?

Reply:
Here is an example. Let's say we have three files: 1.sst, 2.sst, 3.sst. Initially we picked all of them, which make it cover the whole range. Now we eliminated 3 because it is new, but take [1, 2]. According to the logic here, we made `covering_the_whole_range`. With this, we'll update update compaction_end to be the largest key of 2.sst. And in the upper layer, it will call RunManualCompaction() again, just to find that there isn't anything to compact anymore. It's the same thing even if we eliminate 1.sst. It's not wrong but feels intuitive to me.

Reply:
I see, so we should only set `covering_the_whole_range` to false when
1) we filtered out some "new" files in the for-loop
2) in vector `inputs` there are more than 1 "old" files yet to be compacted